torch/fx/partitioner/partitioner.py

```diff
@@ -0,0 +1,183 @@
+                if dependency != 0:
+                    return dependency
+        return 0
+
+    def get_candidates(self):
+        candidates = []
+
+        # TODO: replace following with self.operator_support.is_node_supported()
+        for node in self.module.graph.nodes:
+            if node.op == "call_function":
+                if node.target in self.kernel_registry:
+                    candidates.append(node)
+        return candidates
+
+    def partition(self, candidates: NodeList) -> List[Partition]:
+        # assumptions: nodes in candidate list is sorted in topological order
+        assignment: Dict[Node, int] = {}   # maping from node to partition_id
+        partitions_by_id: Dict[int, Partition] = dict()   # mapping from partition_id to partition
+
+        def assign(node, id):
+            assignment[node] = id
+
+            if id not in partitions_by_id:
+                partitions_by_id[id] = Partition(id=id, nodes=[node])
+            else:
+                partitions_by_id[id].add_node(node)
+
+        def all_equal(iterable):
+            g = groupby(iterable)
+            return next(g, True) and not next(g, False)
+
+        # visit candidates in reversed topological order
+        for node in reversed(candidates):
+
+            user_partitions: Set[Partition] = set()
+            for user_node in node.users:
+                if user_node in assignment:
+                    id = assignment[user_node]
+                    user_partitions.add(partitions_by_id[id])
+                else:
+                    user_partitions.add(Partition(nodes=[user_node]))
+
+            print(node)
+            print('user_partitions', user_partitions)
+
+            # Filter out all the partitions that has dependency on other users
+            # TODO: find a better way to do this, rather than pair-wise comparision
+            user_partitions_list = list(user_partitions)
+            for i in range(len(user_partitions_list)):
+                for j in range(i+1, len(user_partitions_list)):
+                    pi = user_partitions_list[i]
+                    pj = user_partitions_list[j]
+                    dependency = self.__partition_depends_on(pi, pj)
+                    if dependency == 1 and pj in user_partitions:
+                        user_partitions.remove(pj)
+                    elif dependency == -1 and pi in user_partitions:
+                        user_partitions.remove(pi)
+
+            print("user_partitions after filtering", user_partitions)
+
+            # We use the following rules for partition assignment:
+            # 1. If none of the candidates has been assigned to a partition, create a new partition
+            # 2. If all of the candidate has been assigned to the same partition, assign to the same partition
+            # 3. If candidates has been assigned to more then one paritions, assign to the largest partition (by node count)
```

Comment:
I think this heuristic is fine as-is for now.

A quick comment, any reason that we don't want to combine all user partitions? IIUC, at this point there's no dependency among user partitions, so there shouldn't be any reason we can't do that. Our current heuristic (assign the node to the largest partition) looks like is trying to create large partition when possible, we might as well just merge everything.

Reply:
You are right! Thanks for catching this. 
I have updated according to your suggestion.